client/python/client.py

```diff
@@ -0,0 +1,284 @@
+        self._password = password
+
+    def __call__(self, context, callback):
+        b64encoded_auth = base64\
+            .b64encode(bytes(f'{self._username}:{self._password}', 'utf-8'))\
+            .decode('ascii')
+        callback(
+            (('authorization', f'basic {b64encoded_auth}'),),
+            None
+        )
+
+class ArmadaClient:
+    def __init__(self, host: str, port: int,
+        auth_data: AuthData = AuthData(),
+        testing: bool = False):
+
+        self.host = host
+        self.port = port
+        self.testing = testing
+
+        if auth_data.method == AuthMethod.Anonymous:
+            self.channel = grpc.insecure_channel(f'{host}:{port}')
+        elif auth_data.method == AuthMethod.Basic:
+            if self.testing:
+                channel_credentials = grpc.local_channel_credentials()
+            else:
+                # TODO pass root certs, private key, cert chain if this is needed
+                channel_credentials = grpc.ssl_channel_credentials()
+
+            self.channel = grpc.secure_channel(f'{host}:{port}',
+                grpc.composite_channel_credentials(
+                    channel_credentials,
+                    grpc.metadata_call_credentials(
+                        GrpcBasicAuth(auth_data.username, auth_data.password)
+                    )
+                )
+            )
+        elif auth_data.method == AuthMethod.OpenId:
+            pass
+        elif auth_data.method == AuthMethod.Kerberos:
+            pass
+
+        self.submit_stub = submit_pb2_grpc.SubmitStub(self.channel)
+        self.event_stub = event_pb2_grpc.EventStub(self.channel)
+        self.usage_stub = usage_pb2_grpc.UsageStub(self.channel)
+
+    def get_job_events_stream(self, queue, job_set_id, from_message_id=None, watch=False):
+        jsr = event_pb2.JobSetRequest(queue=queue, from_message_id=from_message_id, watch=False)
+        jse = self.event_stub.GetJobSetEvents(queue, job_set_id, jsr)
+        print(jse)
+
+    def submit_jobs(self, queue, job_set_id, job_request_items):
+        request = submit_pb2.JobSubmitRequest(
+            queue=queue,
+            job_set_id=job_set_id,
+            job_request_items=job_request_items
+        )
+        response = self.submit_stub.SubmitJobs(request)
+        return response
+
+    def cancel_jobs(self, queue=None, job_id=None, job_set_id=None):
+        request = submit_pb2.JobCancelRequest(
+            queue=queue, job_id=job_id, job_set_id=job_set_id
+        )
+        response = self.submit_stub.CancelJobs(request)
+        return response
+
+    def reprioritize_jobs(self, new_priority, job_ids=None, job_set_id=None, queue=None):
+        request = submit_pb2.JobReprioritizeRequest(
+            job_ids=job_ids, job_set_id=job_set_id, queue=queue,
+            new_priority=new_priority
+        )
+        response = self.submit_stub.ReprioritizeJobs(request)
+        return response
+
+    def create_queue(self, name, **queue_params):
```

Comment:
Is this the way to do it in Python? Why are we wrapping the auto-generated code like this? Couldn't we just call into the auto-generated code directly?

Reply:
All right.